core/include/ikos/core/domain/memory/value.hpp

```diff
@@ -1495,6 +1551,12 @@ class ValueDomain final
       bool first = true;
 
       for (MemoryLocationRef addr : addrs) {
+        if (is_read_uninitialized(addr, offset, size)) {
+          // No cell found for the base address, so must not be initialized.
+          this->uninit_refine(lhs.var(), Uninitialized::uninitialized());
+          return;
+        }
```

Comment:
I think this is unsound. There are other memory locations that this pointer could point to (`addrs` is a set), so we shouldn't assume here that `lhs` is uninitialized.

For instance:
```c
  int* p = NULL;
  int x = 0;
  if (c) {
    p = malloc(sizeof(int)); // uninitialized
  } else {
    p = &x;
  }
  int y = *p; // initialized if p = &x but unititialized if p = malloc(..)
```

Reply:
@arthaud, True. I'm working on some modifications which I'll push soon. Your example illustrates an issue, namely, that it is possible for a pointer to (potentially) point both to initialized and uninitialized memory. In that situation we can't say that it is "definitely" uninitialized, but we can say it _might_ be  uninitialized. So instead of exiting the loop and returning, as my current code does, the loop should do a join on the "uninitialization state" of all the locations pointed to. So if "initialized" and "uninitialized" are joined, we get top which will give a warning (rather than an error). 

I have one other issue I would like your feedback on before proceeding with another push. My current strategy is to only create cells for writes, or for reads at locations already initialized. So there is an invariant that uninitialized memory has no underlying cells. Do you think that is a reasonable approach? Alternately we could go ahead and create a cell and then store information on each cell as to whether it was initialized. Which approach do you prefer?

Reply:
> I have one other issue I would like your feedback on before proceeding with another push. My current strategy is to only create cells for writes, or for reads at locations already initialized. So there is an invariant that uninitialized memory has no underlying cells. Do you think that is a reasonable approach? Alternately we could go ahead and create a cell and then store information on each cell as to whether it was initialized. Which approach do you prefer?

If I remember correctly, a missing cell should be treated as top (i.e, memory can be anything, initialized or not).
This is because we have to handle writes where we don't know the exact pointer address, see
https://github.com/NASA-SW-VnV/ikos/blob/master/core/include/ikos/core/domain/memory/value.hpp#L1393-L1416
In these cases we might end up removing cells:
https://github.com/NASA-SW-VnV/ikos/blob/master/core/include/ikos/core/domain/memory/value.hpp#L1156

For instance, this is needed here:
```
char buffer[100];
int offset = external_function_returning_any_integer();
buffer[offset] = 0;
```
We don't know where it writes, and we cannot assume that it writes at all locations `0 <= offset < MAX_INT`. So in that case we remove cells.

To conclude, in cases where you know for sure something is uninitialized, you should probably create a single cell covering it and mark is uninitialized. I guess you will need to change how we handle memory allocation (for instance `malloc()` and stack allocations with `ar::Allocate`) to create a cell covering the whole memory and marking it uninitialized?

Reply:
@arthaud, Thanks for the feedback. I will use that approach.

Reply:
@arthaud, one other question. What is the strategy as far as overlapping cells? I'll study the code to try to find out but if you know the answer that will help me. Do you try to keep the cells as a partition of the space, or allow overlapping cells. If you allow overlaps, then when a cell is written, do you try to update the information on the overlapping cells?

Reply:
@arthaud, also, I realize you are no longer working at NASA. Is there someone else who has maintaining IKOS as part of their job and that would be considered heading the IKOS project? Do you use IKOS at your current job, or do you do code reviews, etc. on your own time?

Reply:
@arthaud, regarding when cells are missing and considered TOP, then a read should be considered as possibly uninitialized, shouldn't it (i.e. the top value for initialization), i.e. a warning. Do you agree.

Reply:
> What is the strategy as far as overlapping cells?

I think we don't allow overlapping cells.

> I realize you are no longer working at NASA. Is there someone else who has maintaining IKOS as part of their job and that would be considered heading the IKOS project? Do you use IKOS at your current job, or do you do code reviews, etc. on your own time?

@guillaume66 has been keeping an eye on the project but I have been the main reviewer, on my own time. I do not use IKOS at my current job.

> regarding when cells are missing and considered TOP, then a read should be considered as possibly uninitialized, shouldn't it (i.e. the top value for initialization), i.e. a warning. Do you agree?

Yes, reading TOP means it can be uninitialized and hence should be a warning.
The downside is that this might trigger a lot of false positives when the analysis gets imprecise.